### PR TITLE
DDF-5802 Refactor to enable for split catalog

### PR DESCRIPTION
**Pull request by lamhuy**
#### What does this PR do?
This PR does some refactoring, renaming to set ground for the split catalog provider work to come. No new functionality is added.

1. Rename SolrCatalogProvider to SolrCatalogProviderImpl
2. ~~Add hint on query for GazetteerQueryCatalog~~
3. Refactor SolrMetacardClientImpl.query logic (no functional change)
4. ~~Add new method to SolrClientFactory Interface~~
5. Add clean up steps for itests

This PR is 1/4 of the work of the original large PR https://github.com/codice/ddf/pull/5813

The remaining 3/4 of the work will be on a follow up PR, which are mostly new code and new modules to support the implementation of split catalog provider

#### Who is reviewing it? 
@pklinef 
@Lambeaux 
@rzwiefel 
@bdeining 

#### Ask 2 committers to review/merge the PR and tag them here.
@pklinef 
@bdeining 

#### How should this be tested?
Full CI should do to ensure all regression tests are covered.

#### What are the relevant tickets?
Fixes: #____

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Threat Dragon models
- [x] Update / Add Unit Tests
- [x] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

**Comment by lamhuy** [comment]
build now

File: distribution/ddf-common/src/main/resources-filtered/etc/application-definitions/solr.json [hunk 1]
```diff
@@ -2,6 +2,7 @@
   "name": "catalog-solr-app",
   "description": "The Solr Catalog Provider (SCP) is an implementation of the CatalogProvider interface using Apache Solr ${solr.version} as a data store.\nThe SCP supports extensible metacards, fast/simple contextual searching, indexes xml attributes/CDATA sections/XML text elements, contains full XPath support, works with an embedded local Solr Server, and also works with a standalone Solr Server.::Solr Catalog",
   "bundleLocations": [
+    "mvn:org.codice.thirdparty/jts/${jts.bundle.version}",
```
Comment (bdeining) [inline distribution/ddf-common/src/main/resources-filtered/etc/application-definitions/solr.json:5]: why is this added to the application definition?

File: distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/AbstractIntegrationTest.java [hunk 2]
```diff
@@ -329,9 +326,6 @@ public String toString() {
   public static final DynamicUrl SERVICE_ROOT =
       new DynamicUrl(SECURE_ROOT, HTTPS_PORT, "/services");
 
-  public static final DynamicUrl INSECURE_SERVICE_ROOT =
```
Comment (bdeining) [inline distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/AbstractIntegrationTest.java:332]: I think this is used downstream, is it necessary to remove?

Comment (lamhuy) [reply distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/AbstractIntegrationTest.java:332]: i can put that back.

File: distribution/ddf-common/src/main/resources-filtered/etc/application-definitions/solr.json (hunk 1 above)
Comment (lamhuy) [reply distribution/ddf-common/src/main/resources-filtered/etc/application-definitions/solr.json:5]: to be consistent with how the feature is composed of https://github.com/lamhuy/ddf/blob/master/catalog/solr/catalog-solr-app/src/main/resources/features.xml#L43

File: distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/AbstractIntegrationTest.java (hunk 2 above)
Comment (lamhuy) [reply distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/AbstractIntegrationTest.java:332]: if it is used downstream, should that variable to added to the extend class rather? i did a quick search and don't see its usage tho.

**Comment by lamhuy** [comment]
build now

File: platform/solr/solr-factory-impl/src/main/java/org/codice/solr/factory/impl/SolrCloudClientFactory.java [hunk 3]
```diff
@@ -64,47 +68,157 @@
...
+                .stream()
+                .filter(c -> c.startsWith(alias))
+                .filter(c -> currentAliases.contains(c))
+                .collect(Collectors.toList());
+      }
+    }
+    return Collections.emptyList();
+  }
+
+  private void createAlias(CloudSolrClient client, List<String> collections, String alias) {
```
Comment (Lambeaux) [inline platform/solr/solr-factory-impl/src/main/java/org/codice/solr/factory/impl/SolrCloudClientFactory.java:165]: ❓ Per my knowledge of Elasticsearch, an `alias` is just an alternate name you can use to refer to an index or group of indices in a REST request, but the database itself owns the alias definition as part of configuration. It seems Solr works in a similar fashion: 
https://lucene.apache.org/solr/guide/8_1/aliases.html

This is a lot more manageable than before, so I'm good with it 👍 but I'm still curious about the necessity for having our Solr client manage Alias creation. Can we not assume that we deployed Solr / Zookeeper with the correct configuration up front? Can we not simply reboot the system or refresh bundles between configuration updates?

Comment (lamhuy) [reply platform/solr/solr-factory-impl/src/main/java/org/codice/solr/factory/impl/SolrCloudClientFactory.java:165]: Your observation is correct. DDF does bear the responsibly to setup solr if it is not configured. But if Solr is already configured, either manually or on previous DDF run, then the current or update configuration won't have any impact.

File: platform/solr/solr-factory/src/main/java/org/codice/solr/factory/SolrClientFactory.java [hunk 4]
```diff
@@ -30,5 +31,7 @@
    * @return the newly created {@code SolrClient}
    * @throws IllegalArgumentException if <code>core</code> is <code>null</code>
    */
-  public SolrClient newClient(String core);
+  SolrClient newClient(String core);
+
+  SolrClient newClient(String collection, Map<String, Object> properties);
```
Comment (pklinef) [inline platform/solr/solr-factory/src/main/java/org/codice/solr/factory/SolrClientFactory.java:36]: :pencil2: Add JavaDoc.  You can reference the existing method.

**Comment by lamhuy** [comment]
build now

File: platform/solr/solr-factory-impl/src/main/java/org/codice/solr/factory/impl/SolrCloudClientFactory.java [hunk 5]
```diff
@@ -64,47 +68,157 @@
...
-          "Cannot create SolrCloud client without Zookeeper host list system property [solr.cloud.zookeeper] being set.");
-      throw new IllegalStateException("system property 'solr.cloud.zookeeper' is not configured");
-    }
     LOGGER.debug(
-        "Solr({}): Creating a SolrCloud client using Zookeeper hosts [{}]", core, zookeeperHosts);
+        "Solr({}): Creating a Solr Cloud client with configuration using Zookeeper hosts [{}]",
+        collection,
+        zookeeperHosts);
+
+    String alias = properties == null ? null : (String) properties.getOrDefault(ALIAS_PROP, null);
```
Comment (rzwiefel) [inline platform/solr/solr-factory-impl/src/main/java/org/codice/solr/factory/impl/SolrCloudClientFactory.java:90]: ✏️ Just an opinion, firming up the null contracts with the properties parameter could clean up null checks like this and reduce it down to just the getOrDefault call.

File: platform/solr/solr-factory-impl/src/main/java/org/codice/solr/factory/impl/SolrCloudClientFactory.java [hunk 6]
```diff
@@ -64,47 +68,157 @@
...
-                      return createSolrCloudClient(zookeeperHosts, core);
+                      return createSolrCloudClient(zookeeperHosts, collection, alias);
                     }));
   }
 
   @VisibleForTesting
-  SolrClient createSolrCloudClient(String zookeeperHosts, String collection) {
+  void addCollectionToAlias(String alias, String collection, CloudSolrClient client) {
+    if (StringUtils.isBlank(alias) || StringUtils.isBlank(collection)) {
+      return;
```
Comment (rzwiefel) [inline platform/solr/solr-factory-impl/src/main/java/org/codice/solr/factory/impl/SolrCloudClientFactory.java:105]: ❓ Just confirming, is silently returning (instead of throwing an exception) what we want here if the alias or collection is null?

File: platform/solr/solr-factory-impl/src/main/java/org/codice/solr/factory/impl/SolrCloudClientFactory.java [hunk 7]
```diff
@@ -64,47 +68,157 @@
...
+    CollectionAdminResponse aliasResponse =
+        new CollectionAdminRequest.ListAliases().process(client);
+
+    if (aliasResponse != null) {
+      Map<String, String> aliases = aliasResponse.getAliases();
+      if (aliases != null && aliases.containsKey(alias)) {
+        String[] collections = aliases.get(alias).split(",");
+        return Arrays.asList(collections);
+      }
+    }
```
Comment (rzwiefel) [inline platform/solr/solr-factory-impl/src/main/java/org/codice/solr/factory/impl/SolrCloudClientFactory.java:132]: ✏️ This would be a great place to use Optional if you prefer to have the null checks done for you, eg:
```java
return Optional.ofNullable(aliasResponse)
    .map(CollectionAdminResponse::getAliases)
    .map(a -> a.get(alias))
    .map(ca -> ca.split(",")
    .map(Arrays::asList)
    .orElse(Collections.emptyList());
```

File: platform/solr/solr-factory-impl/src/main/java/org/codice/solr/factory/impl/SolrCloudClientFactory.java [hunk 8]
```diff
@@ -64,47 +68,157 @@
...
+      CollectionAdminResponse response = new CollectionAdminRequest.List().process(client);
+      if (response.getResponse() != null) {
+        List<String> collections = (List<String>) response.getResponse().get("collections");
+        return collections == null
+            ? Collections.emptyList()
+            : collections
+                .stream()
+                .filter(c -> c.startsWith(alias))
+                .filter(c -> currentAliases.contains(c))
+                .collect(Collectors.toList());
```
Comment (rzwiefel) [inline platform/solr/solr-factory-impl/src/main/java/org/codice/solr/factory/impl/SolrCloudClientFactory.java:159]: ✏️ Bummer optionals and exceptions dont play well together or you could probably do more with this. But even the latter half could be Optionaled if you'd prefer
```java
return Optional.ofNullable(response)
    .map(CollectionAdminResponse::getResponse)
    .map(resp -> resp.get("collections"))
    .orElse(Collections.emptyList())
    .stream()
    .filter(c -> c.startsWith(alias))
    .filter(c -> currentAliases.contains(c))
    .collect(Collectors.toList());
```

File: platform/solr/solr-factory-impl/src/main/java/org/codice/solr/factory/impl/HttpSolrClientFactory.java [hunk 9]
```diff
@@ -139,6 +140,12 @@ private static boolean solrCoreExists(SolrClient client, String coreName)
     return new SolrClientAdapter(core, () -> createSolrHttpClient(solrUrl, core, coreUrl));
   }
 
+  @Override
+  public org.codice.solr.client.solrj.SolrClient newClient(
+      String collection, Map<String, Object> properties) {
+    return newClient(collection);
```
[the remark this answers is not in the document]
Comment (lamhuy) [reply platform/solr/solr-factory-impl/src/main/java/org/codice/solr/factory/impl/HttpSolrClientFactory.java:146]: yes, this is the deprecated class anyway. having this method just to satisfy the interface.

File: platform/solr/solr-factory-impl/src/main/java/org/codice/solr/factory/impl/SolrCloudClientFactory.java (hunk 6 above)
Comment (lamhuy) [reply platform/solr/solr-factory-impl/src/main/java/org/codice/solr/factory/impl/SolrCloudClientFactory.java:105]: going to remove the null check here. https://github.com/lamhuy/ddf/blob/preSplit/platform/solr/solr-factory-impl/src/main/java/org/codice/solr/factory/impl/SolrCloudClientFactory.java#L212
making this error benign. It just mean noop.

**Comment by lamhuy** [comment]
build now

**Comment by lamhuy** [comment]
build now

**Comment by lamhuy** [comment]
build now

**Comment by lamhuy** [comment]
build now

**Comment by lamhuy** [comment]
build now
